server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,217 @@
+		r.ShutdownReceiver.Receive(c, more)
+		action = onShutdown
+		r.state = complete
+	})
+
+	_, cancel := workflow.WithCancel(ctx)
+	for {
+		s.Select(ctx)
+		switch action {
+		case onNewRevision:
+			revisionCtx := workflow.WithValue(ctx, internalContext.SHAKey, prRevision.Revision)
+			workflow.GetLogger(revisionCtx).Info("received revision")
+			if process := r.shouldProcessRevision(prRevision); !process {
+				continue
+			}
+			// cancel in progress workflow (if it exists) and start up a new one
+			cancel()
+			revisionCtx, cancel = workflow.WithCancel(revisionCtx)
+			r.state = working
+			r.lastAttemptedRevision = prRevision.Revision
+			workflow.Go(revisionCtx, func(c workflow.Context) {
+				r.processRevision(c, prRevision)
+			})
+		case onShutdown:
+			//todo: maybe optimize by cancelling in progress child workflows
+			workflow.GetLogger(ctx).Info("received shutdown request")
+			return nil
+		}
+	}
+}
+
+func (r *Runner) shouldProcessRevision(prRevision receiver.Revision) bool {
+	// ignore reruns when revision is still in progress
+	if r.lastAttemptedRevision == prRevision.Revision && r.state != waiting {
+		return false
+	}
+	return true
+}
+
+// processRevision handles spinning off child Terraform workflows per root and
+// dealing with any failed policies by reviewing set of approvals
+func (r *Runner) processRevision(ctx workflow.Context, prRevision receiver.Revision) {
+	defer func() {
+		r.state = waiting
+	}()
+	failedPolicies := make(map[string]activities.PolicySet)
+	var futures []workflow.ChildWorkflowFuture
+	var rootInfos []RootInfo
+	for _, root := range prRevision.Roots {
+		ctx = workflow.WithValue(ctx, internalContext.ProjectKey, root.Name)
+		future, rootInfo, err := r.processRoot(ctx, root, prRevision)
+		if err != nil {
+			continue
+		}
+		futures = append(futures, future)
+		rootInfos = append(rootInfos, rootInfo)
+	}
+	for i, future := range futures {
+		failedRootPolicies, err := r.awaitWorkflow(ctx, future, rootInfos[i])
+		if err != nil {
+			continue
+		}
+		// consolidate failures across all roots
+		// policy sets are identical so multiple roots can fail the same policy without issue
+		for _, failedPolicy := range failedRootPolicies {
+			failedPolicies[failedPolicy.Name] = failedPolicy
+		}
+	}
+	// TODO: check for policy failures
+}
+
+func (r *Runner) processRoot(ctx workflow.Context, root terraformActivities.Root, prRevision receiver.Revision) (workflow.ChildWorkflowFuture, RootInfo, error) {
+	id, err := sideeffect.GenerateUUID(ctx)
+	if err != nil {
+		workflow.GetLogger(workflow.WithValue(ctx, internalContext.ErrKey, err)).Error("generating uuid")
+		// choosing to not fail workflow and let it continue to exist
+		// until PR close/timeout
+		return nil, RootInfo{}, err
+	}
+	rootInfo := RootInfo{
+		ID: id,
+		Commit: github.Commit{
+			Revision: prRevision.Revision,
+		},
+		Root: root,
+		Repo: prRevision.Repo,
+	}
```

Comment:
weird to be building something here if you're going to return it.  Why not just build it outside of this function and pass it in?